File: crack.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Database Client 优化版解锁脚本
特点：
1. 自动搜索 VSCode 插件目录
2. 不依赖特定文件名 Hash，支持多版本
3. 移除无效的 Localhost 重定向，专注于逻辑覆盖
"""

import os
import re
import sys
import shutil
import platform
from pathlib import Path

# 配置部分
TARGET_EXTENSION_NAME = "cweijan.vscode-"  # 插件ID部分匹配
MOCK_USER_DATA = """{
    id: "cracked_by_script",
    email: "vip@cracked.com",
    username: "VIP_User",
    expireTime: 4102444800000,
    isPremium: true,
    license: "unlimited"
}"""

class PatchManager:
    def __init__(self):
        self.extension_dir = self._find_extension_dir()
        
    def _find_extension_dir(self):
        """尝试自动查找 VSCode 扩展目录"""
        home = Path.home()
        system = platform.system()
        
        # 标准 VSCode 扩展路径
        possible_paths = [
            home / ".vscode" / "extensions",
            home / ".vscode-server" / "extensions", # Remote SSH
        ]
        
        print("正在搜索插件目录...")
        target_dir = None
        
        for base_path in possible_paths:
            if not base_path.exists():
                continue
            # 查找匹配名称的文件夹（处理版本号不同的情况）
            for d in base_path.iterdir():
                if d.is_dir() and d.name.startswith(TARGET_EXTENSION_NAME):
                    target_dir = d
                    break
            if target_dir:
                break
        
        # 如果自动查找失败，尝试使用当前目录
        if not target_dir:
            current_dir = Path(os.getcwd())
            if (current_dir / "package.json").exists() and (current_dir / "out").exists():
                target_dir = current_dir
        
        if not target_dir:
            print(f"❌ 未找到插件目录: {TARGET_EXTENSION_NAME}")
            print("请将此脚本放置在插件根目录下运行，或手动指定路径。")
            sys.exit(1)
            
        print(f"✅ 定位到插件目录: {target_dir}")
        return target_dir

    def backup_file(self, file_path: Path):
        """创建备份，如果已存在备份则跳过"""
        backup_path = file_path.with_suffix(file_path.suffix + '.bak')
        if not backup_path.exists():
            try:
                shutil.copy2(file_path, backup_path)
                print(f"   已备份: {file_path.name}")
            except Exception as e:
                print(f"   ⚠️ 备份失败: {e}")

    def patch_content(self, content, rules):
        """应用一系列正则替换规则"""
        modified_content = content
        count = 0
        for pattern, replacement, desc in rules:
            if re.search(pattern, modified_content):
                modified_content = re.sub(pattern, replacement, modified_content)
                count += 1
                # print(f"      应用规则: {desc}") 
        return modified_content, count

    def process_extension_js(self):
        """处理主进程文件 extension.js"""
        file_path = self.extension_dir / "out" / "extension.js"
        if not file_path.exists():
            print("⚠️ 未找到 out/extension.js，跳过后端补丁")
            return

        print(f"正在处理: {file_path.name}")
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            self.backup_file(file_path)
            
            rules = [
                # 1. 强制静态方法返回 True
                (r'static\s+isPremium\s*\(\s*\)\s*\{[\s\S]*?\}', 'static isPremium() { return true; }', "isPremium = true"),
                (r'static\s+isPay\s*\([^)]*\)\s*\{[\s\S]*?\}', 'static isPay(e) { return true; }', "isPay = true"),
                (r'static\s+isExpire\s*\([^)]*\)\s*\{[\s\S]*?\}', 'static isExpire(e) { return false; }', "isExpire = false"),
                
                # 2. 注入模拟用户
                (r'static\s+getUser\s*\(\s*\)\s*\{[\s\S]*?\}', f'static getUser() {{ return {MOCK_USER_DATA}; }}', "Mock User"),
                
                # 3. 绕过网络验证 (不修改 URL，直接修改调用结果)
                # 查找类似 checkLicense() { ... } 的异步函数并短路
                (r'async\s+checkLicense\s*\([^)]*\)\s*\{[\s\S]*?\}', 'async checkLicense() { return true; }', "Bypass checkLicense"),
                (r'async\s+verifyLicense\s*\([^)]*\)\s*\{[\s\S]*?\}', 'async verifyLicense() { return true; }', "Bypass verifyLicense"),
            ]
            
            new_content, count = self.patch_content(content, rules)
            
            if count > 0:
                with open(file_path, 'w', encoding='utf-8') as f:
                    f.write(new_content)
                print(f"   ✅ 成功应用 {count} 处修改")
            else:
                print("   ⚠️ 未找到匹配特征，可能已修改或版本不支持")
                
        except Exception as e:
            print(f"   ❌ 处理出错: {e}")

    def process_webview_assets(self):
        """处理 Webview 资源文件"""
        assets_dir = self.extension_dir / "out" / "webview" / "assets"
        if not assets_dir.exists():
            print("⚠️ 未找到 assets 目录，跳过前端补丁")
            return

        print(f"正在扫描 assets 目录 ({len(list(assets_dir.glob('*.js')))} 个 JS 文件)...")
        
        # 通用规则，应用于所有 JS 文件
        common_rules = [
            # 强制前端判断为已付费
            (r'(!|\!)([a-zA-Z0-9_]+)\.isPay', 'false', "Force !isPay -> false"), 
            (r'([a-zA-Z0-9_]+)\.isPay', 'true', "Force isPay -> true"),
            # 移除连接限制文字
            (r'Database 5/3', 'Database ∞', "Remove DB limit text"),
            (r'Other 3/3', 'Other ∞', "Remove Other limit text"),
            # 隐藏 Premium 徽章 (通过修改 hidden 属性或 value)
            (r'value:(["\'])Premium Only\1', 'value:"",hidden:true', "Hide Premium Label 1"),
            (r'value=(["\'])Premium Only\1', 'value="",hidden=true', "Hide Premium Label 2"),
        ]

        # 特征规则：根据文件内容特征来决定是否应用特定补丁，而不是根据文件名
        specific_rules = [
            {
                "signature": "pay.connectNotice", # 包含这个字符串的文件通常是连接管理页面
                "rules": [
                    (r'innerHTML:.\.\$t\(`pay\.connectNotice`\)', 'innerHTML:""', "Clear connect notice"),
                ]
            },
            {
                "signature": "pricing", # 可能涉及价格显示的逻辑
                "rules": [
                    (r'"pricing"\s*:\s*"Trial"', '"pricing":"Free"', "Trial -> Free"),
                ]
            }
        ]

        patched_count = 0
        for js_file in assets_dir.glob('*.js'):
            if js_file.name.endswith('.bak'): continue
            
            try:
                with open(js_file, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                original_len = len(content)
                new_content = content
                
                # 1. 应用通用规则
                new_content, c_count = self.patch_content(new_content, common_rules)
                
                # 2. 应用基于特征的规则
                for spec in specific_rules:
                    if spec["signature"] in new_content:
                        new_content, s_count = self.patch_content(new_content, spec["rules"])
                        c_count += s_count

                if len(new_content) != original_len or c_count > 0:
                    self.backup_file(js_file)
                    with open(js_file, 'w', encoding='utf-8') as f:
                        f.write(new_content)
                    print(f"   ✅ 已修补: {js_file.name}")
                    patched_count += 1
                    
            except Exception as e:
                print(f"   ❌ 处理 {js_file.name} 出错: {e}")
        
        print(f"前端资源处理完成，共修改 {patched_count} 个文件")

    def process_package_json(self):
        """修改 package.json"""
        pkg_path = self.extension_dir / "package.json"
        if not pkg_path.exists(): return
        
        print("正在检查 package.json...")
        try:
            with open(pkg_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            if '"pricing": "Trial"' in content:
                self.backup_file(pkg_path)
                content = content.replace('"pricing": "Trial"', '"pricing": "Free"')
                with open(pkg_path, 'w', encoding='utf-8') as f:
                    f.write(content)
                print("   ✅ 已修改 pricing 模式")
            else:
                print("   无需修改或未找到 pricing 字段")
        except Exception as e:
            print(f"   ❌ 读取 package.json 失败: {e}")

    def run(self):
        print("🚀 开始执行 Database Client 解锁脚本 (优化版)")
        print("-" * 50)
        self.process_extension_js()
        print("-" * 50)
        self.process_webview_assets()
        print("-" * 50)
        self.process_package_json()
        print("-" * 50)
        print("🎉 完成！请重启 VSCode。")
        print("💡 提示: 如果之前打开过 Database Client，请按 F1 -> 'Developer: Reload Window' 刷新。")

if __name__ == '__main__':
    # 检查权限
    try:
        PatchManager().run()
    except PermissionError:
        print("\n❌ 错误: 权限不足。")
        print("请使用管理员权限 (sudo/Administrator) 运行此脚本。")

File: crack.js
/**
 * Database Client 去除付费限制脚本
 * 
 * 该脚本修改相关文件，使插件的付费功能可用
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// 获取扩展目录路径
const extensionDir = __dirname;
const extensionJsPath = path.join(extensionDir, 'out', 'extension.js');

console.log('系统类型:', os.platform());
console.log('扩展目录路径:', extensionDir);
console.log('extension.js 路径:', extensionJsPath);

// 检查文件是否存在
if (!fs.existsSync(extensionJsPath)) {
  console.error(`错误: 找不到文件 ${extensionJsPath}`);
  console.log('请确认您是否将脚本放在正确的扩展目录中');
  process.exit(1);
}

console.log('开始修改文件以移除付费限制...');

try {
  // 读取extension.js文件
  let extensionJs = fs.readFileSync(extensionJsPath, 'utf8');
  console.log(`成功读取extension.js文件，大小: ${extensionJs.length} 字节`);

  // 备份原始文件
  const backupPath = path.join(extensionDir, 'out', 'extension.js.bak');
  if (!fs.existsSync(backupPath)) {
    fs.writeFileSync(backupPath, extensionJs);
    console.log(`已创建备份文件: ${backupPath}`);
  }

  // 1. 修改isPremium函数 - 始终返回true
  let isPremiumRegex = /static\s+isPremium\s*\(\s*\)\s*\{[\s\S]*?\}/;
  if (isPremiumRegex.test(extensionJs)) {
    extensionJs = extensionJs.replace(isPremiumRegex, 'static isPremium() { return true; }');
    console.log('成功修改 isPremium 函数');
  } else {
    console.log('未找到 isPremium 函数，尝试其他模式...');
    // 尝试其他可能的模式
    isPremiumRegex = /isPremium\s*\(\s*\)\s*\{[\s\S]*?\}/;
    if (isPremiumRegex.test(extensionJs)) {
      extensionJs = extensionJs.replace(isPremiumRegex, 'isPremium() { return true; }');
      console.log('成功修改 isPremium 函数（替代模式）');
    } else {
      console.log('警告: 无法找到 isPremium 函数');
    }
  }

  // 2. 修改isExpire函数 - 始终返回false
  let isExpireRegex = /static isExpire\([^)]*\) \{[\s\S]*?\}/;
  if (isExpireRegex.test(extensionJs)) {
    extensionJs = extensionJs.replace(isExpireRegex, 'static isExpire(e) { return false; }');
    console.log('成功修改 isExpire 函数');
  } else {
    console.log('未找到 isExpire 函数');
  }

  // 3. 修改isPay函数 - 始终返回true
  let isPayRegex = /static isPay\([^)]*\) \{[\s\S]*?return[\s\S]*?\}/;
  if (isPayRegex.test(extensionJs)) {
    extensionJs = extensionJs.replace(isPayRegex, 'static isPay(e) { return true; }');
    console.log('成功修改 isPay 函数');
  } else {
    console.log('未找到 isPay 函数');
  }

  // 4. 修改getUser函数 - 返回一个虚拟的付费用户信息
  let getUserRegex = /static getUser\(\) \{[\s\S]*?\}/;
  if (getUserRegex.test(extensionJs)) {
    extensionJs = extensionJs.replace(getUserRegex, `static getUser() { 
      return { 
        id: "cracked", 
        email: "premium@example.com", 
        username: "premium_user", 
        expireTime: new Date(new Date().getTime() + 365 * 24 * 60 * 60 * 1000).getTime(),
        isPremium: true,
        license: "premium"
      };
    }`);
    console.log('成功修改 getUser 函数');
  } else {
    console.log('未找到 getUser 函数');
  }

  // 5. 查找并修改许可验证相关的API请求函数
  // 封锁网络请求验证
  extensionJs = extensionJs.replace(
    /(https?:\/\/|['"](https?:\/\/))database-client\.com\/api\//g, 
    '$1localhost:1234/api/'
  );
  extensionJs = extensionJs.replace(
    /(https?:\/\/|['"](https?:\/\/))database-client\.com\/#\/console\//g, 
    '$1localhost:1234/#/console/'
  );

  // 6. 修改任何可能用于验证的函数
  extensionJs = extensionJs.replace(
    /async checkLicense\([^)]*\)\s*\{[\s\S]*?\}/g,
    'async checkLicense() { return true; }'
  );
  
  extensionJs = extensionJs.replace(
    /async verifyLicense\([^)]*\)\s*\{[\s\S]*?\}/g,
    'async verifyLicense() { return true; }'
  );
  
  extensionJs = extensionJs.replace(
    /async activateLicense\([^)]*\)\s*\{[\s\S]*?\}/g,
    'async activateLicense() { return true; }'
  );

  // 保存前检查文件权限
  try {
    fs.accessSync(path.dirname(extensionJsPath), fs.constants.W_OK);
  } catch (err) {
    console.error(`错误: 没有写入权限 ${path.dirname(extensionJsPath)}`);
    console.log('请尝试使用管理员/sudo权限运行此脚本');
    process.exit(1);
  }

  // 保存修改后的文件
  try {
    fs.writeFileSync(extensionJsPath, extensionJs);
    console.log('成功保存修改后的extension.js文件');
  } catch (err) {
    console.error('保存文件时出错:', err);
    console.log('请尝试使用管理员/sudo权限运行此脚本');
    process.exit(1);
  }

  // 修改webview中的相关文件
  const webviewAssetsDir = path.join(extensionDir, 'out', 'webview', 'assets');
  if (fs.existsSync(webviewAssetsDir)) {
    console.log(`找到webview资源目录: ${webviewAssetsDir}`);
    
    // 列出目录中的所有文件
    console.log('目录中的文件:');
    try {
      const files = fs.readdirSync(webviewAssetsDir);
      files.forEach(file => console.log(`- ${file}`));
    } catch (err) {
      console.error('读取目录内容时出错:', err);
    }
    
    // 需要重点修改的文件模式
    const priorityPatterns = [
      /^connect-.*\.js$/,  // 连接界面
      /^app-.*\.js$/,      // 应用核心文件
      /^Result-.*\.js$/,   // 结果显示
      /^Main-.*\.js$/,     // 主界面
      /^coreStore-.*\.js$/, // 核心存储
      /^Plan-.*\.js$/      // 付费计划相关
    ];

    // 查找目录中所有的JS文件
    let allJsFiles = [];
    try {
      allJsFiles = fs.readdirSync(webviewAssetsDir)
        .filter(file => file.endsWith('.js') && !file.endsWith('.bak'));
      console.log(`找到 ${allJsFiles.length} 个JS文件`);
    } catch (err) {
      console.error('读取assets目录失败:', err);
    }

    // 处理所有JS文件
    for (const fileName of allJsFiles) {
      const filePath = path.join(webviewAssetsDir, fileName);
      const isPriority = priorityPatterns.some(pattern => pattern.test(fileName));
      
      if (fs.existsSync(filePath)) {
        try {
          console.log(`正在修改 ${fileName}${isPriority ? ' (优先文件)' : ''}...`);
          let fileContent = fs.readFileSync(filePath, 'utf8');
          
          // 创建备份
          const backupFilePath = filePath + '.bak';
          if (!fs.existsSync(backupFilePath)) {
            fs.writeFileSync(backupFilePath, fileContent);
          }
          
          // 1. 替换isPay相关检查
          fileContent = fileContent.replace(/(!|\!)([a-zA-Z0-9_]+)\.isPay/g, 'false');
          fileContent = fileContent.replace(/([a-zA-Z0-9_]+)\.isPay/g, 'true');
          fileContent = fileContent.replace(/C\(t\)\.isPay/g, 'true');
          fileContent = fileContent.replace(/W\(a\)\.isPay/g, 'true');
          
          // 2. 替换 hidden 属性，确保 Premium Only 标记被隐藏
          fileContent = fileContent.replace(/hidden:!([a-zA-Z0-9_]+)\.isPay/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:([a-zA-Z0-9_]+)\.isPay/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:!C\(t\)\.isPay/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:C\(t\)\.isPay/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:!d\.value/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:!f\.value/g, 'hidden:true');
          fileContent = fileContent.replace(/hidden:!a\.value/g, 'hidden:true');
          
          // 3. 处理连接限制的显示文本
          fileContent = fileContent.replace(/Database 5\/3/g, 'Database ∞/∞');
          fileContent = fileContent.replace(/Other 3\/3/g, 'Other ∞/∞');
          fileContent = fileContent.replace(/连接限制: Database 5\/3, Other 3\/3/g, '连接无限制');
          
          // 4. 特别处理"Premium Only"标签文本
          fileContent = fileContent.replace(/value:(["'])Premium Only\1/g, 'value:$1$1');
          fileContent = fileContent.replace(/value=(["'])Premium Only\1/g, 'value=$1$1');
          
          // 5. 特别处理 connect-BtFhhlKZ.js 文件
          if (fileName === 'connect-BtFhhlKZ.js') {
            // 定位并处理Premium Only标签
            fileContent = fileContent.replace(/\{"class":"text-base green mt-1","value":"Premium Only",/g, 
                                           '{"class":"text-base green mt-1","value":"", "hidden":true,');
            
            // 处理连接限制提示
            fileContent = fileContent.replace(/innerHTML:t\.\$t\(`pay\.connectNotice`\)/g, 
                                           'innerHTML:""');
            
            // 确保工作区范围选项可用
            fileContent = fileContent.replace(/disabled:!f\.value/g, 'disabled:false');
            fileContent = fileContent.replace(/disabled:!d\.value/g, 'disabled:false');
            fileContent = fileContent.replace(/disabled:!a\.value/g, 'disabled:false');
          }
          
          // 6. 特别处理 app-rfvkh6uB.js 文件
          if (fileName === 'app-rfvkh6uB.js') {
            fileContent = fileContent.replace(/\{"value":"Premium Only",/g, 
                                           '{"value":"", "hidden":true,');
          }
          
          // 7. 移除许可限制提示
          fileContent = fileContent.replace(/闭源扩展，完整功能需要Premium许可/g, '');
          fileContent = fileContent.replace(/Premium许可将移除此徽章/g, '');
          fileContent = fileContent.replace(/需要Premium/g, '');
          
          // 8. 阻止验证请求 - 修改API请求URL
          fileContent = fileContent.replace(
            /(https?:\/\/|['"](https?:\/\/))database-client\.com\/api\//g, 
            '$1localhost:1234/api/'
          );
          
          // 9. 修改动态验证的状态检查
          // 找到可能进行许可状态检查的代码
          if (fileName === 'coreStore-m1eBg2Tl.js') {
            // 修改许可状态验证逻辑
            fileContent = fileContent.replace(
              /isPay\s*\([^)]*\)\s*\{[^}]*\}/g,
              'isPay(){return true}'
            );
            
            // 检查并修改许可状态的初始化
            fileContent = fileContent.replace(
              /status\s*:\s*{\s*code\s*:\s*[0-9]/g,
              'status:{code:0'
            );
          }
          
          // 10. 特别处理Plan-BHeFVyRe.js付费计划相关文件
          if (fileName === 'Plan-BHeFVyRe.js') {
            // 替换付费状态检查
            fileContent = fileContent.replace(
              /I\.isPay/g, 
              'true'
            );
          }

          // 保存修改后的文件
          fs.writeFileSync(filePath, fileContent);
          console.log(`成功修改 ${fileName}`);
        } catch (err) {
          console.error(`修改 ${fileName} 时出错:`, err);
        }
      } else {
        console.log(`文件 ${fileName} 不存在`);
      }
    }
  }

  // 修改package.json中的pricing
  const packageJsonPath = path.join(extensionDir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      let packageJson = fs.readFileSync(packageJsonPath, 'utf8');
      // 备份原始文件
      const packageJsonBackup = packageJsonPath + '.bak';
      if (!fs.existsSync(packageJsonBackup)) {
        fs.writeFileSync(packageJsonBackup, packageJson);
      }
      
      // 替换pricing字段
      packageJson = packageJson.replace(/"pricing"\s*:\s*"Trial"/, '"pricing": "Free"');
      
      fs.writeFileSync(packageJsonPath, packageJson);
      console.log('成功修改 package.json 文件');
    } catch (err) {
      console.error('修改 package.json 时出错:', err);
    }
  }

  // 创建空的模拟验证API响应文件
  try {
    const mockApiDir = path.join(extensionDir, 'mock_api');
    if (!fs.existsSync(mockApiDir)) {
      fs.mkdirSync(mockApiDir, { recursive: true });
    }
    
    // 创建模拟验证响应
    const mockResponsePath = path.join(mockApiDir, 'mock_license_response.json');
    const mockData = {
      code: 0,
      data: {
        id: "cracked",
        email: "premium@example.com",
        username: "premium_user",
        expireTime: new Date(new Date().getTime() + 365 * 24 * 60 * 60 * 1000).getTime(),
        isPremium: true,
        license: "premium"
      },
      message: "success"
    };
    
    fs.writeFileSync(mockResponsePath, JSON.stringify(mockData, null, 2));
    console.log('创建了模拟验证响应文件');
  } catch (err) {
    console.error('创建模拟响应文件时出错:', err);
  }

  console.log('所有修改已完成，请重启VSCode以使更改生效。');
  console.log('提示: 如果重启后功能仍受限，请在VSCode中按F1，然后执行 "Developer: Reload Window" 命令，或完全关闭VSCode后重新打开。');
  console.log('注意: 如果使用后仍然出现Premium Only，可能需要在网络设置中断开网络连接后再使用VSCode，或者通过hosts文件将database-client.com指向127.0.0.1');
} catch (err) {
  console.error('修改过程中出错:', err);
} 
